pkg/apiutil/serverapi/middleware.go

```diff
@@ -40,32 +41,76 @@ const (
 
 var initHTTPClientOnce sync.Once
 
+// dialClient used to dail http request.
+var dialClient = &http.Client{
+	Transport: &http.Transport{
+		DisableKeepAlives: true,
+	},
+}
+
+type runtimeServiceAuth struct {
+	s *server.Server
+}
+
+// NewRuntimeServiceAuth checks if the path is invalid.
+func NewRuntimeServiceAuth(s *server.Server) *runtimeServiceAuth {
+	return &runtimeServiceAuth{s: s}
+}
+
+func (h *runtimeServiceAuth) ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	if IsServiceAllowed(h.s, r.RequestURI) {
+		next(w, r)
+		return
+	}
+
+	http.Error(w, "no service", http.StatusServiceUnavailable)
+}
+
+// IsServiceAllowed checks the service through the path.
+func IsServiceAllowed(s *server.Server, path string) bool {
```

Comment:
This function is called after routing, and its return value should be unchanged after startup, and can be determined at NewRuntimeServiceAuth. Can the configuration change at runtime?

In addition, assuming that there is `foo` in RuntimeServices, a `/pd/apis/keyvisual/v1/foo` like path can maliciously bypass the check. Can we use APIGroup.Name directly to check.

Reply:
Yes, Server option can be changed in the runtime. `/pd/apis/keyvisual/v1/foo` can not bypass, what do you mean?